transports/quic/src/certificate.rs

```diff
@@ -0,0 +1,235 @@
+        let contents = reader.next().read_bytes()?;
+        match &**oid.components() {
+            LIBP2P_OID => verify_libp2p_extension(&contents, certificate_key).map(Some),
+            _ => Ok(None),
+        }
+    })
+}
+
+fn verify_libp2p_extension(
+    extension: &[u8],
+    certificate_key: &[u8],
+) -> yasna::ASN1Result<identity::PublicKey> {
+    yasna::parse_der(extension, |reader| {
+        reader.read_sequence(|mut reader| {
+            let public_key = read_bitvec(&mut reader)?;
+            let signature = read_bitvec(&mut reader)?;
+            let public_key = identity::PublicKey::from_protobuf_encoding(&public_key)
+                .map_err(|_| yasna::ASN1Error::new(yasna::ASN1ErrorKind::Invalid))?;
+            let mut v = Vec::with_capacity(LIBP2P_SIGNING_PREFIX_LENGTH + certificate_key.len());
+            v.extend_from_slice(&LIBP2P_SIGNING_PREFIX[..]);
+            v.extend_from_slice(certificate_key);
+            if public_key.verify(&v, &signature) {
+                Ok(public_key)
+            } else {
+                Err(yasna::ASN1Error::new(yasna::ASN1ErrorKind::Invalid))
+            }
+        })
+    })
+}
+
+fn parse_certificate(certificate: &[u8]) -> yasna::ASN1Result<identity::PublicKey> {
+    trace!("parsing certificate");
+    yasna::parse_der(certificate, |reader| {
+        reader.read_sequence(|reader| {
+            let key = parse_tbscertificate(reader.next())?;
+            // skip algorithm
+            reader.next().read_der()?;
+            // skip signature
+            reader.next().read_der()?;
+            Ok(key)
+        })
+    })
+}
+
+fn parse_tbscertificate(reader: yasna::BERReader) -> yasna::ASN1Result<identity::PublicKey> {
+    trace!("parsing TBScertificate");
+    reader.read_sequence(|reader| {
+        // Check the X.509 version
+        if reader.next().read_der()? != [160, 3, 2, 1, 2] {
+            return Err(yasna::ASN1Error::new(yasna::ASN1ErrorKind::Invalid));
+        }
+        // Skip the serial number
+        reader.next().read_der()?;
+        // Skip the signature algorithm
+        reader.next().read_der()?;
+        // Skip the issuer
+        reader.next().read_der()?;
+        // Skip validity
+        reader.next().read_der()?;
+        // Skip subject
+        reader.next().read_der()?;
+        trace!("reading subjectPublicKeyInfo");
+        let key = reader.next().read_sequence(|mut reader| {
+            // Skip the subject key algorithm
+            reader.next().read_der()?;
+            trace!("reading subject key");
+            read_bitvec(&mut reader)
+        })?;
+        trace!("reading extensions");
+        parse_x509_extensions(reader, &key)
+    })
+}
+
+/// The name is a misnomer. We don’t bother checking if the certificate is actually well-formed.
+/// We just check that its public key is suitably signed.
+pub fn verify_libp2p_certificate(certificate: &[u8]) -> Result<libp2p_core::PeerId, webpki::Error> {
```

Comment:
Since the function name is a misnomer, what about `extract_libp2p_peerid` instead, for example?